### PR TITLE
[Metrics] AUC/AUROC class interface

### LINKED ISSUE #4293: error when using auroc
I may be using `auroc` incorrectly, but this seems like a bug, which is reproducable using the boring model example for tensors with ndim > 1

## 🐛 Bug

Using the auroc() metric results in an error:
```RuntimeError: Boolean value of Tensor with more than one value is ambiguous```

## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

https://colab.research.google.com/drive/1Dc5cVu8FZN1C2ZnH4ZH-ouG2II1dt4FE?usp=sharing

### Expected behavior

It should work as is?
this line 

https://github.com/PyTorchLightning/pytorch-lightning/blob/96785b99df878ebc45bc455eb7902a6dd4661139/pytorch_lightning/metrics/functional/classification.py#L892

should use torch.any()?

### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0): 1.6.0
 - OS (e.g., Linux): mac
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.8
 - CUDA/cuDNN version: none
 - GPU models and configuration:  none
 - Any other relevant information: none

### Additional context

also reproducible here:

Screenshot 2020-10-21 at 22 07 25

### LINKED ISSUE #4547: Class version of AUROC metric
## 🚀 Feature

Class version of AUROC metric following the v1.x.x standard, so we can do:
```python
auroc = AUROC(num_classes=4)
auroc(pred, target)
auroc.compute()
```

### Motivation

Class based Metrics can be nicer to use in the PyTorch Lightning workflow, so why not add one for AUROC?

### Pitch

AUROC will automatically use the correct functional function, `auroc` or `multiclass_auroc`, based on whether `num_classes == 2` or `num_classes > 2`, respectively.
Not sure why `auroc` or `multiclass_auroc` are 2 separate functions?

### Alternatives

Implement the AUROC Metric yourself.

### Additional context

**Considerations:**
How to handle `update()`? While it's true that most likely the whole dataset will include all classes in `pred` and `targets`, this cannot be assumed for a single batch.

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #4293
Fixes #4547 

Implementation of class based interface for AUC and AUROC and unification with their respective functional counterpart.
Brings a number of improvements to AUROC also:
* `auroc` and `multiclass_auroc` are combined into `auroc`
* support for multilabel (not sure if we already did this, but now we test for it and state it in the docs)
* support for multi dim tensors
* support for different averaging in multiclass setting

This should be one of last PRs before the hole metric package is unified.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/metrics/classification/test_auc.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def sk_auc(x, y, reorder=False):
+    x = x.flatten()
+    y = y.flatten()
+    if reorder:
+        idx = np.argsort(x, kind='stable')
+        x = x[idx]
+        y = y[idx]
+        print('sk', x, y, idx)
```
Comment (marload) [inline tests/metrics/classification/test_auc.py:23]: I think it's better to erase debug print
```python
print('sk', x, y, idx)
```

File: pytorch_lightning/metrics/utils.py [hunk 2]
```diff
@@ -264,3 +264,20 @@ def class_reduce(
...
+    See this discussion: https://discuss.pytorch.org/t/is-torch-sort-stable/20714
+    """
+    if x.ndim > 1:
+        raise ValueError('Stable sort only works on 1d tensors')
+    n = x.numel()
+    if 2049 - n > 0:
+        x_max = x.max()
+        x_pad = torch.cat([x, (x_max + 1) * torch.ones(2049 - n, dtype=x.dtype, device=x.device)], 0)
+    x_sort = x_pad.sort()
+    return x_sort.values[:n], x_sort.indices[:n]
```
Comment (marload) [inline pytorch_lightning/metrics/utils.py:283]: I think `2049` should be variable.

File: pytorch_lightning/metrics/functional/auroc.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            elif average == 'weighted':
+                if mode == 'multi-label':
+                    support = torch.sum(target, dim=0)
+                else:
+                    support = torch.bincount(target.flatten(), minlength=num_classes)
+                return torch.sum(torch.stack(auc_scores) * support / support.sum())
+            else:
+                allowed_average = (None, 'macro', 'weighted')
+                raise ValueError('Argument `average` expected to be one of the following:'
+                                 f' {allowed_average} but got {average}')
```
Comment (marload) [inline pytorch_lightning/metrics/functional/auroc.py:96]: `if - if - if - None` pattern is better than `if - elif - elif` in Early Return Pattern

**Review by marload — CHANGES_REQUESTED** [review]
Great Work 👍 
But I have some request changes.

File: tests/metrics/classification/test_auc.py (hunk 1 above)
Comment (SkafteNicki) [reply tests/metrics/classification/test_auc.py]: you are right, stupid leftover from debugging :]

File: pytorch_lightning/metrics/utils.py (hunk 2 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/utils.py:284]: Done

File: pytorch_lightning/metrics/functional/auroc.py (hunk 3 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/auroc.py]: I agree, have changed it now :]

**Review by teddykoker — APPROVED** [review]
Brilliant! Thank you so much @SkafteNicki

File: pytorch_lightning/metrics/classification/auc.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    Args:
+        reorder: AUC expects its first input to be sorted. If this is not the case,
+            setting this argument to ``True`` will use a stable sorting algorithm to
+            sort the input in decending order
+        compute_on_step:
+            Forward only calls ``update()`` and return None if this is set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step. default: False
```
Comment (Borda) [inline pytorch_lightning/metrics/classification/auc.py:38]:
```suggestion
            Forward only calls ``update()`` and return None if this is set to False.
        dist_sync_on_step:
            Synchronize metric state across processes at each ``forward()``
            before returning the value at the step.
```

File: pytorch_lightning/metrics/classification/auc.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            ' For large datasets this may lead to large memory footprint.'
+        )
+
+    def update(self, x: torch.Tensor, y: torch.Tensor):
+        """
+        Update state with predictions and targets.
+
+        Args:
+            preds: Predictions from model (probabilities, or labels)
+            target: Ground truth labels
```
Comment (Borda) [inline pytorch_lightning/metrics/classification/auc.py:76]:
```suggestion
            x: Predictions from model (probabilities, or labels)
            y: Ground truth labels
```

File: pytorch_lightning/metrics/classification/auroc.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+              where each class is weighted by their support (accounts for class imbalance)
+            - ``None`` computes and returns the metric per class
+        max_fpr:
+            If not ``None``, calculates standardized partial AUC over the
+            range [0, max_fpr]. Should be a float between 0 and 1.
+        compute_on_step:
+            Forward only calls ``update()`` and return None if this is set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step. default: False
```
Comment (Borda) [inline pytorch_lightning/metrics/classification/auroc.py:61]:
```suggestion
            before returning the value at the step.
```

File: pytorch_lightning/metrics/classification/auroc.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            raise ValueError('Argument `average` expected to be one of the following:'
+                             f' {allowed_average} but got {average}')
+
+        if self.max_fpr is not None:
+            if (not isinstance(max_fpr, float) and 0 < max_fpr <= 1):
+                raise ValueError(f"`max_fpr` should be a float in range (0, 1], got: {max_fpr}")
+
+            if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
+                raise RuntimeError('`max_fpr` argument requires `torch.bucketize` which'
+                                   ' is not available below PyTorch version 1.6')
```
Comment (Borda) [inline pytorch_lightning/metrics/classification/auroc.py:123]:
```suggestion
                raise RuntimeError(
                    '`max_fpr` argument requires `torch.bucketize` which is not available below PyTorch version 1.6'
                )
```

File: pytorch_lightning/metrics/functional/auc.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    if reorder:
+        x, x_idx = _stable_1d_sort(x)
+        y = y[x_idx]
+    dx = x[1:] - x[:-1]
+    if (dx < 0).any():
+        if (dx <= 0).all():
+            direction = -1.
+        else:
+            raise ValueError("The `x` tensor is neither increasing or decreasing."
+                             "Try setting the reorder argument to `True`.")
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/auc.py:41]:
```suggestion
            raise ValueError(
                "The `x` tensor is neither increasing or decreasing."
                " Try setting the reorder argument to `True`."
            )
```

File: pytorch_lightning/metrics/functional/auroc.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        # max_fpr parameter is only support for binary
+        if mode != 'binary':
+            raise ValueError("Partial AUC computation not available in "
+                             "multilabel/multiclass setting, 'max_fpr' must be"
+                             f" set to `None`, received `{max_fpr=}`.")
+
+    if mode == 'binary':
+        num_classes = 1
+
+    if mode == 'multi-label':
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/auroc.py:68]:
```suggestion
    elif mode == 'multi-label':
```

File: pytorch_lightning/metrics/functional/auroc.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+    if max_fpr is None or max_fpr == 1:
+        if num_classes != 1:
+            # calculate auc scores per class
+            auc_scores = [auc(x, y) for x, y in zip(fpr, tpr)]
+
+            # calculate average
+            if average is None:
+                return auc_scores
+            if average == 'macro':
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/auroc.py:85]:
```suggestion
            elif average == 'macro':
```

File: pytorch_lightning/metrics/functional/auroc.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                return torch.mean(torch.stack(auc_scores))
+            if average == 'weighted':
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/auroc.py:87]:
```suggestion
            elif average == 'weighted':
```

File: pytorch_lightning/metrics/functional/classification.py [hunk 12]
```diff
@@ -605,6 +619,12 @@ def multiclass_auroc(
         >>> multiclass_auroc(pred, target, num_classes=4)
         tensor(0.6667)
     """
+    rank_zero_warn(
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:622]: add test for this call

**Review by marload — APPROVED** [review]

File: pytorch_lightning/metrics/functional/auroc.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        fpr, tpr, _ = roc(preds, target, num_classes, pos_label, sample_weights)
+
+    # calculate standard roc auc score
+    if max_fpr is None or max_fpr == 1:
+        if num_classes != 1:
+            # calculate auc scores per class
+            auc_scores = [auc(x, y) for x, y in zip(fpr, tpr)]
+
+            # calculate average
+            if average is None:
```
Comment (tchaton) [inline pytorch_lightning/metrics/functional/auroc.py:86]: Could we move mode into enum ?

**Review by tchaton — APPROVED** [review]
LGTM ! Great work !

File: pytorch_lightning/metrics/functional/auroc.py (hunk 10 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/auroc.py:88]: i think it is considered better practise using if-if-if insted of if-elif-elif when each block call `return` in the end

File: pytorch_lightning/metrics/functional/auroc.py (hunk 13 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/auroc.py:86]: what do you mean? should `average` be a `Enum`?

File: pytorch_lightning/metrics/functional/auroc.py (hunk 10 above)
Comment (Borda) [reply pytorch_lightning/metrics/functional/auroc.py:88]: well the diff is also in logic/processing speed, you clearly see that only one of these is possible, compare to if-if-if you still need to validate each if (mentally while reading and computer while running it)

File: pytorch_lightning/metrics/functional/classification.py [hunk 14]
```diff
@@ -500,6 +504,12 @@ def auc(
         >>> auc(x, y)
         tensor(4.)
     """
+    rank_zero_warn(
+        "This `auc` was deprecated in v1.2.0 in favor of"
+        " `pytorch_lightning.metrics.functional.auc import auc`."
+        " It will be removed in v1.4.0", DeprecationWarning
+    )
+
     dx = x[1:] - x[:-1]
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:513]: can we internally call the new function so we do not need to maintain the code in 2 places till 1.4?

File: pytorch_lightning/metrics/functional/classification.py [hunk 15]
```diff
@@ -568,6 +582,12 @@ def auroc(
         >>> auroc(x, y)
         tensor(0.5000)
     """
+    rank_zero_warn(
+        "This `auroc` was deprecated in v1.2.0 in favor of"
+        " `pytorch_lightning.metrics.functional.auroc import auroc`."
+        " It will be removed in v1.4.0", DeprecationWarning
+    )
+
     if any(target > 1):
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:591]: can we internally call the new function so we do not need to maintain the code in 2 places till 1.4?

File: pytorch_lightning/metrics/functional/classification.py [hunk 16]
```diff
@@ -630,6 +654,12 @@ def multiclass_auroc(
         >>> multiclass_auroc(pred, target, num_classes=4)
         tensor(0.6667)
     """
+    rank_zero_warn(
+        "This `multiclass_auroc` was deprecated in v1.2.0 in favor of"
+        " `pytorch_lightning.metrics.functional.auroc import auroc`."
+        " It will be removed in v1.4.0", DeprecationWarning
+    )
+
     if not torch.allclose(pred.sum(dim=1), torch.tensor(1.0)):
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/classification.py:663]: can we internally call the new function so we do not need to maintain the code in 2 places till 1.4?

File: pytorch_lightning/metrics/utils.py [hunk 17]
```diff
@@ -264,3 +264,21 @@ def class_reduce(
...
+
+
+def _stable_1d_sort(x: torch):
+    """
+    Stable sort of 1d tensors. Pytorch defaults to a stable sorting algorithm
+    if number of elements are larger than 2048. This function pads the tensors,
+    makes the sort and returns the sorted array (with the padding removed)
+    See this discussion: https://discuss.pytorch.org/t/is-torch-sort-stable/20714
+    """
+    N = 2049
```
Comment (Borda) [inline pytorch_lightning/metrics/utils.py:276]: can we move this as a parameter with default?

**Comment by Borda** [comment]
@SkafteNicki can we in the deprecated function call new func, just to lover room for potential bugs... 🐰

File: pytorch_lightning/metrics/functional/auroc.py (hunk 13 above)
Comment (Borda) [reply pytorch_lightning/metrics/functional/auroc.py:86]: I think he means to define average types as LightEnum, for example, as we have https://github.com/PyTorchLightning/pytorch-lightning/blob/671887fd9bd37bc44086549b18f2ee2de40f71e5/pytorch_lightning/trainer/states.py#L39-L49

File: pytorch_lightning/metrics/classification/auroc.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+        Args:
+            preds: Predictions from model (probabilities, or labels)
+            target: Ground truth labels
+        """
+        preds, target, mode = _auroc_update(preds, target)
+
+        self.preds.append(preds)
+        self.target.append(target)
+        self.mode = mode
```
Comment (tchaton) [inline pytorch_lightning/metrics/classification/auroc.py:147]: Should the mode be saved in a list too ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/metrics/utils.py (hunk 17 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/utils.py:276]: Done :]

File: pytorch_lightning/metrics/functional/classification.py (hunk 14 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/classification.py]: Done :]

File: pytorch_lightning/metrics/functional/classification.py (hunk 15 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/classification.py]: Done :]

File: pytorch_lightning/metrics/functional/classification.py (hunk 16 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/classification.py:635]: Done :]

File: pytorch_lightning/metrics/classification/auroc.py (hunk 18 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/classification/auroc.py:151]: Thanks for the suggestion. So instead of saving it to a list, I instead included a check that `self.mode==mode`, which makes sure that the mode of the data (binary, multi-class, multi-label) did not change between batches (which it really should not).

File: pytorch_lightning/metrics/functional/auroc.py [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+                               " is not available below PyTorch version 1.6")
+
+        # max_fpr parameter is only support for binary
+        if mode != 'binary':
+            raise ValueError(f"Partial AUC computation not available in "
+                             f"multilabel/multiclass setting, 'max_fpr' must be"
+                             f" set to `None`, received `{max_fpr}`.")
+
+    # calculate fpr, tpr
+    if mode == 'multi-label':
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/auroc.py:78]: lest do it in another PR, but also move this to Enum :]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/metrics/functional/auroc.py (hunk 19 above)
Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/auroc.py:78]: Agree :]
I think multiple places in metrics could be converted into Enum

Comment (Borda) [reply pytorch_lightning/metrics/functional/auroc.py:78]: cool, co lets create an issue for and mark is good-first-issue :]

Comment (SkafteNicki) [reply pytorch_lightning/metrics/functional/auroc.py:78]: I do that :]

**Comment by Borda** [comment]
@SkafteNicki mind resolve conflicts? 🐰
